### PR TITLE
the Big Xenomorphs update

### PR DESCRIPTION
**Pull request by DeityLink**
nests

Timelapse: https://www.youtube.com/watch?v=A8Sy6OO_xUQ
- Fixed the Adminbus being destroyable by Corrosive Acid.
- "Simple_Animals" xenomorphs (the ones that aren't player controlled) now have a brand new AI! They will spread weeds, break lights near them, and queens will place nests and eggs. They will automatically drag monkeys and incapacitated humans into those nests. Incidentally this means that xenobio will be able to create real xenos if they manage to gold slime an alien queen into existence!
- NPC xenos no longer instantly break walls, instead they cover them with hyper acid that destroys them over 5 to 10 seconds. Not a very impactful change, but looks nice.
- Mobs stuck onto xeno nests now visually appear partially inside nest.
- Xeno weeds now progressively die out over time when the weed nodes are destroyed.
- Xeno eggs now check for nearby potential victims as soon as they become grown, rather than having to wait for a valid human to walk nearby. This way they can jump onto victims that had been stuck onto nests before they were fully grown.
- You can attack burst xeno egg to fully remove them now.
- All "hot" weapons now deal full damage to xeno eggs and weeds, previously only lit welder did.
- Plasma cutters are now "hot" (which means you can light up your cigarette with one)
- Getting stuck onto a xeno nest will stop you from bleeding out (from external wounds), as well as keep you supplied on a bit of inaprovaline, leporazine, and dexaline, sensibly improving your chances of survival (as well as the chances for a successful facehugger rape).
- Facehuggers will now target monkeys as well as humans by themselves.
- monkeys now appear bellow "bedsheets"-layer items when lying down.

alienstation13

### CODE REVIEW

**Comment by Ephx1** [comment]
> Incidentally this means that xenobio will be able to create real xenos if they manage to gold slime an alien queen into existence

Oh god fuck off

**Comment by CptWad** [comment]
AHAHHAHAHA YES
xenobio making xenos might end poorly but otherwise YES

**Comment by Exxion** [comment]
I don't know if you would consider this too snowflakey, but you might consider having the nests inject vox with nitrogen rather than inaprovaline, as inaprov is toxic to vox
1/20 as much if you want to have it last the same amount of time

**Comment by CptWad** [comment]
or hell just invent a chemical called "Stabilizine" or something have it stop bleeding simulate having proper air to breath and prevent further oxy damage like inprovaline

**Comment by Exxion** [comment]
OOH OOH WAIT I KNOW I KNOW
Use stoxin2
It acts as a stronger inaprovaline that DOES work on vox but also puts you to sleep

**Comment by DeityLink** [comment]
@CptWad @Exxion 

> or hell just invent a chemical called "Stabilizine" or something have it stop bleeding, simulate having proper air to breath and prevent further oxy damage like inprovaline

I like that idea, I'm stealing it

**Comment by CptWad** [comment]
ur welcum

**Comment by ghost** [comment]
srejuvenate?

**Comment by DeityLink** [comment]
No, I don't want the nests to fully heal them, nor do I want it to put them asleep.

**Comment by unid15** [comment]
:ok_hand: spicy

**Comment by Exxion** [comment]
srejuvenate (stoxin2) doesn't actually heal, despite its description
But it does put them to sleep so that part is true

**Comment by Brolaire** [comment]
This looks hype as shit, blessed be the children of Deity, codermin prophety
Did you just slap a modified Spider AI on them or did you do it from scratch

**Comment by DeityLink** [comment]
I don't really see what I could have re-used from the spider AI. Often spiders will disregard actual mobs in order to go break lights, I didn't want that to happen with my aliums so they automatically check for adjacent lights to break when they wander around, but don't purposefully move toward lights.

They both have the basic hostile simple animal AI (wandering around, lunging after targets) and all the stuff about placing structures and gathering preys is from scratch.

**Comment by NigglyWiggly** [comment]
> Getting stuck onto a xeno nest will stop you from bleeding out (from external wounds), as well as keep you supplied on a bit of inaprovaline, leporazine, and dexaline, sensibly improving your chances of survival (as well as the chances for a successful facehugger rape).

Oh jesus christ man

**Comment by Ephx1** [comment]
Really though, remove them from gold cores or make it so the simple mob xeno doesn't lay eggs.
You don't know how fucking retarded of a change that is.

It's like retarded xeno miner syndrome jammed to max.

**Comment by ihadtoregisterforthis** [comment]
This truly is a different take on making xenos a gamemode.

**Comment by Boxe** [comment]
does this fix the stupid thing where facehuggers rip off RIG helmets

or was that fixed already

**Comment by Ephx1** [comment]
@Boxe 
it's a _feature_

**Comment by DeityLink** [comment]
@ihadtoregisterforthis
I know right? People have wanted to breed xenos in a controlled environment for a long time.

@Ephx1 

> make it so the simple mob xeno doesn't lay eggs.

That's like....the entire point of this PR that you're asking me to roll back you know?

Creating player-controlled xenomorphs in xenobio is just as risky as creating player-controlled slimes. Both can wreck up a station. Xenobiologists will now have the ability to prepare a pen to contain them, have them fight each others, etc, it will be fun!

**Comment by ghost** [comment]
@Ephx1 Gold slimes can summon spider nurses and queens, which can then take the corpses scattered around and make eggs, causing a spider pandemic.

This is just spiders but they actually basebuild

**Comment by Exxion** [comment]
Every station except Box (and maybe Taxi, I don't remember) has a containment pen in xenobio at roundstart that just requires a wrench and an engineer's ID to set up to be basically indestructible

**Comment by Ephx1** [comment]
Spiders don't have player control behind them.
Spiders can't ventcrawl.
Spiders can't figure out a good escape route.

If you do breed xenos they always get out, ALWAYS. It's the reason why I want to fucking stab miners that bring back a hugged monkey. Also your "Indestructible" pen is incredibly flawed, the piping is its main flaw, plasma piping can be melted and there's fuck all you can do.

**Comment by DeityLink** [comment]
@Ephx1 If a xenobiologist wants to destroy the station he really doesn't need player controlled xenos to do it. 

Think of it as an experiment. Don't you just want to see what's gonna happen? I for one can't wait for xenobiologist to run to medbay no only to heal slime clone damage but now also xeno embryos.

**Comment by Ephx1** [comment]
No, he doesn't, that's not the point. In the similar way a non-antag miner brings a hugger back, he's going to fuck up the station regardless.

Nothing is going to happen other than them getting out. What the fuck are you going to do with xenos trapped in a pen, stare at them? No, they're not going to fight, that's retarded.

Really, if you want this to be a thing make some research value based off of them, otherwise it's just a hazard that's going to happen more frequently.

**Comment by DeityLink** [comment]
You make a very good point, one could remove the entirety of toxins and bomb testing with that reasoning, but I'll let it slide because you're giving me new ideas.

I could make xenos produce something that the scientists might want to acquire, or add xeno-only explosive implants that scientists use on pregnant monkeys to ensure that they have a degree of failsafe on rogue xenos.

Anything to bring some fresh air to this game can't be bad. If there are flaws, we'll dig around them.

**Comment by Ephx1** [comment]
Look towards the comics if you want more ideas based on them, xenos have absurd forms and whacky shit that scientists did to them.

**Comment by ghost** [comment]
If we were to go towards the comics and extended universe, the evolution trees would have to be reworked (Dependent on species rather than being able to go drone, hunter, or sentinel off the bat, making a queen would require a praetorian egg and facehugger, praetorian then evolves into a queen, etc.)

**Comment by DeityLink** [comment]
I've never read alien comics, guess now I have a reason to.

Wait, what

pig

Either way, most people here are excited about the prospect of xenobio-bred player controlled xenos so I say let's give it a shot. We'll make further changes based on the feedback we get. For instance, maybe xenobiology will now have an extra electrified pen, that'll be for mappers to add.

**Comment by CptWad** [comment]
Hell its not like this is un-do able. If this becomes a problem then its easily fixed, hell we can build more off of it.

**Comment by ghost** [comment]
I was hoping to build off the AI for other need-to-be-managed-else-shitstorm mob types, so the sooner this all gets livetested, the cooler

**Comment by NigglyWiggly** [comment]
> Either way, most people here are excited about the prospect of xenobio-bred player controlled xenos so I say let's give it a shot

**Comment by DeityLink** [comment]
@Skullyton 
Hey hold on, I'm not done having fun coding AIs! I saw that we have simple animal cultists now right? There no way I can let anyone else code their AI!

**Comment by ghost** [comment]
@DeityLink I wanted to adapt what you've done, with the basebuilding and dragging of corpses and such, you can have that humanoid AI stuff

**Comment by DeityLink** [comment]
well my code was optimized for those xenos in particular. Another kind of humanoid AI might benefit better from another way of writing this code.

**Comment by DeityLink** [comment]
Done a few final changes. Weed nodes will pulsate every 30 seconds and make their children check whether they can expand again (in case a door opened/a wall broke, for example)

Just like webs, weed will jam airlocks open.

Aliens will now try to open doors near them.

**Comment by CptWad** [comment]
>Just like webs, weed will jam airlocks open.

OH BOY

**Comment by ihadtoregisterforthis** [comment]
And resin doors won't get jammed, right?

**Comment by NigglyWiggly** [comment]
> Just like webs, weed will jam airlocks open.

Why would you do this

**Comment by DeityLink** [comment]
@ihadtoregisterforthis
Haha holy shit this is why I shouldn't code just before going to sleep. That would have been hilarious but thanks for avoiding a catastrophe. Now the weeds make sure the door isn't made of resin.

@NigglyWiggly
Their AI needs a non-obstructed path to be able to reliably bring preys back to the nests. Making the doors they come across as their base expand stay open kinda ensures that a bit.

You only need to destroy the weeds under the door to be able to close it again.

**Comment by gbasood** [comment]
> Xeno eggs now check for nearby potential victims as soon as they become grown, rather than having to wait for a valid human to walk nearby. This way they can jump onto victims that had been stuck onto nests before they were fully grown.

What? When a facehugger comes out of its egg it auto-leaps on nearby targets because it triggers Crossed() when it's moved out of the egg.

**Comment by DeityLink** [comment]
@gbasood Crossed only triggers when you move on the same tile the atom is. If you move on an adjacent tile it's HasProximity() that will trigger. When the egg bursts, the egg itself checks for valid targets adjacent and stucks the facehugger on them. If there are no valid targets adjacent, the facehugger just stands still, even if there is a valid target 2 tiles away. Now I make the facehugger actually call findtarget() after the egg bursted (which looks for valid targets up top 4 tiles away, and makes the facehugger moves toward them).

**Comment by SonixApache** [comment]
How effective are flamethrowers?

**Comment by DeityLink** [comment]
er, flamethrowers are "a little" effective to remove weeds, but as weapon they need a serious rework. They don't "feel" like flamethrowers at all, more like pouring a bucket of gasoline on the floor.

**Comment by PJB3005** [comment]
@DeityLink even after @ShadowMech88's rework?

**Comment by DeityLink** [comment]
If there was any rework to flamethrowers in the past 3 months then disregard what i just said. I have no idea.

**Comment by Ephx1** [comment]
@DeityLink 
#8318

**Comment by gbasood** [comment]
@DeityLink I can tell you, with absolute certainty, that it currently triggers Crossed() when the facehugger is moved out of the egg onto the tile, though.

The reason it does is because it calls forceMove() which triggers Crossed().
If you manually set the loc of the facehugger it wouldn't trigger it though

File: code/game/objects/effects/aliens.dm [hunk 1]
```diff
@@ -200,35 +205,60 @@
...
-	for(var/obj/effect/alien/weeds/W in connected_weeds)
-		W.linked_node = null
-	..()
 
 /obj/effect/alien/weeds/node/New()
 	connected_weeds = new()
 	..(src.loc, src)
 
+/obj/effect/alien/weeds/node/Destroy()
+	for(var/obj/effect/alien/weeds/W in connected_weeds)
```
Comment (clusterfack) [inline code/game/objects/effects/aliens.dm]: Nobody thinks to null connected_weeds?

File: code/game/objects/effects/aliens.dm [hunk 2]
```diff
@@ -200,35 +205,60 @@
...
+
+/obj/effect/alien/weeds/node/proc/heartbeat()
+	flick("weednode-heartbeat",src)
+
+	for(var/obj/effect/alien/weeds/W in connected_weeds)
+		if(W.health == 15)
+			spawn()
+				W.Life()
+	spawn(HEARTBEAT_RATE)
+		heartbeat()
```
Comment (clusterfack) [inline code/game/objects/effects/aliens.dm]: Try something more like
if(!gcDestroyed) .()

File: code/game/objects/effects/aliens.dm [hunk 3]
```diff
@@ -200,35 +205,60 @@
...
 
+/obj/effect/alien/weeds/proc/Life()
+	health = 15
 	//set background = 1
 	var/turf/U = get_turf(src)
 /*
```
Comment (clusterfack) [inline code/game/objects/effects/aliens.dm]: Please remove this commented out section it adds nothing

File: code/modules/mob/mob.dm [hunk 4]
```diff
@@ -1489,7 +1489,7 @@ var/list/slot_equipment_priority = list( \
 		canmove = has_limbs
 
 	if(lying)
-		if(ishuman(src))
+		if(ishuman(src) || ismonkey(src))
```
Comment (clusterfack) [inline code/modules/mob/mob.dm]: iscarbon perhaps? Why are these singled out

File: code/modules/mob/living/carbon/human/life/handle_chemicals_in_body.dm [hunk 5]
```diff
@@ -5,6 +5,10 @@
 
 	if(reagents)
 
+		if(locked_to && istype(locked_to,/obj/structure/bed/nest))//alien nest will keep their victims alive for a bit longer.
```
Comment (clusterfack) [inline code/modules/mob/living/carbon/human/life/handle_chemicals_in_body.dm]: Have nests do you this you silly silly person, don't modify general procs for interactions with specific objects

File: code/game/objects/structures/stool_bed_chair_nest/alien_nests.dm [hunk 6]
```diff
@@ -57,7 +68,9 @@
 
 /obj/structure/bed/nest/proc/healthcheck()
 	if(health <= 0)
-		density = 0
+		if(locked_atoms.len)
+			var/mob/M = locked_atoms[1]
```
Comment (clusterfack) [inline code/game/objects/structures/stool_bed_chair_nest/alien_nests.dm]: This should be covered by qdel() no?

Comment (PJB3005) [inline code/game/objects/structures/stool_bed_chair_nest/alien_nests.dm]: Yeah it is.

File: code/modules/mob/mob.dm (hunk 4 above)
Comment (PJB3005) [inline code/modules/mob/mob.dm]: Slimes are carbons IIRC.

**Comment by DeityLink** [comment]
@gbasood yes but crossed will only matter in this case if you are standing Right on top of the egg. Also the egg's Crossed calls HasProximity() (which checks in a 1 tile radius).

File: code/game/objects/effects/aliens.dm (hunk 2 above)
Comment (DeityLink) [inline code/game/objects/effects/aliens.dm]: The point of heartbeat() is to allow the previously spawned weeds to keep expanding if a wall has been destroyed/a door has been openned. (also it has a flick animation so it looks like the node is alive every 30 seconds) What would if(!gcDestroyed) do here?

File: code/modules/mob/mob.dm (hunk 4 above)
Comment (DeityLink) [inline code/modules/mob/mob.dm]: Yep, don't want them to regard slimes (or worse, other player-controlled aliens) as valid prey.

File: code/game/objects/effects/aliens.dm (hunk 2 above)
Comment (clusterfack) [inline code/game/objects/effects/aliens.dm]: Keep it from stopping the proc from garbage collecting

File: code/modules/mob/mob.dm (hunk 4 above)
Comment (clusterfack) [inline code/modules/mob/mob.dm]: This is canmove, you are using canmove for targetting?

File: code/modules/mob/living/carbon/human/life/handle_chemicals_in_body.dm (hunk 5 above)
Comment (DeityLink) [inline code/modules/mob/living/carbon/human/life/handle_chemicals_in_body.dm]: I was under the impression that adding a Life/process tick to nests would be even worse but I guess I didn't think at the time that I could just have the proc stop when nobody is stuck on the nest.

**Comment by DeityLink** [comment]
I'll take care of the issues when I get back.

**Comment by gbasood** [comment]
Oh, you're checking for pathfinding, that's different. Nevermind then.

**Comment by Probe1** [comment]
<3

File: code/game/objects/effects/aliens.dm (hunk 1 above)
Comment (DeityLink) [inline code/game/objects/effects/aliens.dm]: I'm adding "connected_weeds = null" in /obj/effect/alien/weeds/node/Destroy()

File: code/modules/mob/mob.dm (hunk 4 above)
Comment (DeityLink) [inline code/modules/mob/mob.dm]: oh, nvm then

**Comment by DeityLink** [comment]
Done

File: code/game/objects/structures/stool_bed_chair_nest/alien_nests.dm [hunk 7]
```diff
@@ -66,13 +66,36 @@
 	user.delayNextAttack(10)
 	healthcheck()
 
+/obj/structure/bed/nest/bullet_act(var/obj/item/projectile/Proj)
+	health -= Proj.damage
+	..()
+	healthcheck()
+	return
```
Comment (PJB3005) [inline code/game/objects/structures/stool_bed_chair_nest/alien_nests.dm]: Trailing return.

**Comment by gbasood** [comment]
100% test o clock

**Comment by DeityLink** [comment]
Ah shit when I woke up this morning I rememberd that I had forgotten to have the nests call stabilize() so they currently don't heal their prisonners at all. I'm making a hotfix right now then.

**Comment by Probe1** [comment]
unce unce unce unce unce unce unce unce unce unce unce unce :eggplant:

**Pull request merged**